xtask/src/swift.rs

```diff
@@ -0,0 +1,97 @@
+use std::{env, fs, path::PathBuf};
+
+use clap::{Args, Subcommand};
+use serde::Deserialize;
+use xshell::{cmd, pushd};
+
+use crate::Result;
+
+#[derive(Args)]
+pub struct SwiftArgs {
+    #[clap(subcommand)]
+    cmd: Option<SwiftCommand>,
+}
+
+#[derive(Subcommand)]
+enum SwiftCommand {
+    /// Build as a static lib.
+    BuildLibrary,
+    /// Build as an XCFramework.
+    BuildFramework,
+}
+
+impl SwiftArgs {
+    pub fn run(self) -> Result<()> {
+        let _p = pushd(&workspace_root()?)?;
+
+        match self.cmd {
+            Some(cmd) => match cmd {
+                SwiftCommand::BuildLibrary => build_library(),
+                SwiftCommand::BuildFramework => build_xcframework(),
+            },
+            None => build_library(),
+        }
+    }
+}
+
+fn build_library() -> Result<()> {
+    println!("Running debug library build.");
+
+    let root_directory = workspace_root()?;
+    let target_directory = root_directory.join("target");
+    let ffi_directory = root_directory.join("bindings/apple/generated/matrix_sdk_ffi");
+    let swift_directory = root_directory.join("bindings/apple/generated/swift");
+
+    let release_type = "debug";
+    let static_lib_filename = "libmatrix_sdk_ffi.a";
+
+    fs::create_dir_all(ffi_directory.as_path())?;
+    fs::create_dir_all(swift_directory.as_path())?;
+
+    cmd!("cargo build -p matrix-sdk-ffi").run()?;
+
+    fs::rename(
+        target_directory.join(release_type).join(static_lib_filename),
+        ffi_directory.join(static_lib_filename),
+    )?;
+
+    cmd!(
+        "uniffi-bindgen generate
+    			--language swift
+    			--lib-file {ffi_directory}/{static_lib_filename}
+    			--out-dir {ffi_directory}
+				{root_directory}/bindings/matrix-sdk-ffi/src/api.udl"
+    )
+    .run()?;
+
+    let module_map_file = ffi_directory.join("module.modulemap");
+    if module_map_file.exists() {
+        fs::remove_file(module_map_file.as_path())?;
+    }
+
+    // TODO: Find the modulemap in the ffi directory.
+    fs::rename(ffi_directory.join("matrix_sdk_ffiFFI.modulemap"), module_map_file)?;
+    // TODO: Move all swift files.
+    fs::rename(
+        ffi_directory.join("matrix_sdk_ffi.swift"),
+        swift_directory.join("matrix_sdk_ffi.swift"),
+    )?;
+
+    Ok(())
+}
+
+fn build_xcframework() -> Result<()> {
+    println!("XCFramework not yet implemented.");
+    Ok(())
+}
+
+fn workspace_root() -> Result<PathBuf> {
```

Comment:
I've added copy of this function as both `ci.rs` and `fixup.rs` also have a copy, but wondering whether it would be better to use a shared implementation across all tasks?

Reply:
sure, just move it and reuse it :)

Reply:
I think I've done that the right way 🤞